src/engine/renderer/gl_shader.cpp

```diff
@@ -1139,6 +1184,120 @@ void GLShaderManager::CompileAndLinkGPUShaderProgram( GLShader *shader, shaderPr
 	LinkProgram( program->program );
 }
 
+// This will generate all the extra code for material system shaders
+std::string GLShaderManager::ShaderPostProcess( GLShader *shader, const std::string& shaderText, const bool isVertexShader ) {
```

Comment:
Why not writing a real `.glsl` file to be used to prefix the GLSL shaders with `fragmentInline`/`vertexInline` mechanism, using `ifdef` to select various code path and set defines based on the tests you need?

A good example of a different shaders using a single source but different “inline” parts is generic and generic2D, that both uses `generic_fp.glsl` while using different `compileMacros` and `vertexInlines`. See `GLShader_generic2D` and `GLShader_generic` in `gl_shader.cpp`.

Reply:
> Why not writing a real `.glsl` file to be used to prefix the GLSL shaders with `fragmentInline`/`vertexInline` mechanism, using `ifdef` to select various code path and set defines based on the tests you need? _Edit:_ This is basically an “include” mechanism.

Because I don't want to add a macro for almost every uniform. It would also mean moving all those uniform definitions to a new file for each shader, which is bad. It would also mean having to align memory for every shader manually, and with any change to uniforms too.

Also, having 2 different inlines isn't gonna work here.

Reply:
> Why not writing a real `.glsl` file to be used to prefix the GLSL shaders with `fragmentInline`/`vertexInline` mechanism, using `ifdef` to select various code path and set defines based on the tests you need? _Edit:_ This is basically an conditional “include” mechanism.

Two more issues with that is you'd have to keep all uniforms in the exact same order everywhere, manually, and it probably won't work for textures.

Reply:
I suppose this is somewhat the case now since I did have to put some things into separate shader code files.